### PR TITLE
support for threading in two different ways for esmf

### PR DESCRIPTION
**Pull request by jedwards4b**
Adds threading support for nuopc in two ways with the introduction of a new xml variable in cmeps 
ESMF_AWARE_THREADING.   When this variable is False or None the threading implementation is the same as that in mct, this is the default.  When ESMF_AWARE_THREADING is True the driver uses all available mpi tasks on a given node and component threads are launched using pthreads.  This allows for each component to specify its thread count independent of the other components.  It also changes the definition of ROOTPE.  The pelayout was changed to add PSTRID and to document the value of ESMF_AWARE_THREADING and how that value affects the ROOTPE variable.  

Test suite: scripts_regression_tests.py + manual tests of pelayout script 
Test baseline: 
Test namelist changes: 
Test status: bit for bit

Fixes [CIME Github issue #]
Depends on https://github.com/ESCOMP/CMEPS/pull/144
 
User interface changes?: 

Update gh-pages html (Y/N)?:

Code review:

### CODE REVIEW

File: config/cesm/machines/config_machines.xml [hunk 1]
```diff
@@ -472,6 +472,7 @@ This allows using a different mpirun command to launch unit tests
       <arguments>
         <arg name="labelstdout">-p "%g:"</arg>
         <arg name="num_tasks"> -np {{ total_tasks }}</arg>
+        <arg name="zthreadplacement"> omplace -tm open64 -vv</arg>
```
Comment (jedwards4b) [inline config/cesm/machines/config_machines.xml:475]: Needed when ESMF_AWARE_THREADING is off, not desirable when it is on. leaving it on for now.

File: config/cesm/machines/config_machines.xml [hunk 2]
```diff
@@ -633,16 +634,18 @@ This allows using a different mpirun command to launch unit tests
...
     <environment_variables unit_testing="true">
       <env name="MPI_USE_ARRAY">false</env>
     </environment_variables>
     <environment_variables queue="share">
       <env name="TMPDIR">/glade/scratch/$USER</env>
-<!--      <env name="MPI_USE_ARRAY">false</env> -->
     </environment_variables>
     <resource_limits>
-      <resource name="RLIMIT_STACK"> -1</resource>
+      <resource name="RLIMIT_STACK"> 80000000</resource>
```
Comment (jedwards4b) [inline config/cesm/machines/config_machines.xml:648]: -1 value does not work with pthreads, the value here was chosen via trial and error.  We need to find a better way to set this.

File: scripts/lib/CIME/get_timing.py [hunk 3]
```diff
@@ -90,7 +90,8 @@ def _gettime2_mct(self, heading_padded):
     def _gettime2_nuopc(self):
         self.nprocs = 0
         self.ncount = 0
-        expression = re.compile(r'\s*\MED:\s*\(med_phases_profile\)\s+(\d+)\s+(\d+)')
+#        expression = re.compile(r'\s*\MED:\s*\(med_phases_profile\)\s+(\d+)\s+(\d+)')
```
Comment (jedwards4b) [inline scripts/lib/CIME/get_timing.py:93]: unrelated bug fix for timing output

**Review by jgfouca — APPROVED** [review]

**Review by jgfouca — APPROVED** [review]

File: scripts/lib/CIME/get_timing.py [hunk 4]
```diff
@@ -459,11 +459,10 @@ def _getTiming(self, inst=0):
...
         tmaxr = 0
         if tmax > 0:
-            print ("HERE adays {} tmax {}".format(adays, tmax))
             tmaxr = adays*86400.0/(tmax*365.0)
 
         self.write("\n")
-        self.write("  total pes active           : {} \n".format(totalpes*maxthrds*smt_factor ))
+        self.write("  total pes active           : {} \n".format(totalpes*smt_factor ))
         self.write("  mpi tasks per node               : {} \n".format(max_mpitasks_per_node))
         self.write("  pe count for cost estimate : {} \n".format(pecost))
```
Comment (amametjanov) [inline scripts/lib/CIME/get_timing.py:467]: Since you're here anyways, could you consider updating this?
1. Indentation of `:` for "mpi tasks per node" line
2. It's hard to estimate the cost of a job: e.g. "charge factor" at NERSC Cori. I think it'd be useful to simplify and write
```
  total nodes         : N (num_nodes)
  mpi tasks per node  : M (MAX_MPITASKS_PER_NODE)
  threads per mpi task: T (max_threads)
  pes per node        : MxT
  total pes active    : MxTxN
```
i.e. remove `smt_factor` from all PE calculations. It's hard to account for SMT/hyper-threading correctly without introducing additional XML variable for actual number of cores per node: e.g. OLCF Summit node with 42 cores per node capable of SMT[1,2,4].

**Review by fischer-ncar — APPROVED** [review]

File: scripts/lib/CIME/get_timing.py (hunk 4 above)
Comment (jedwards4b) [reply scripts/lib/CIME/get_timing.py]: I have addressed the indentation.  The smt_factor change is more extensive and not something I feel can be handled as a minor addition.  I would be happy to review a new PR if you would like to submit one.

Comment (amametjanov) [reply scripts/lib/CIME/get_timing.py]: One-line change:
```
diff --git a/scripts/lib/CIME/get_timing.py b/scripts/lib/CIME/get_timing.py
index 68c2e7769..6e197ead4 100644
--- a/scripts/lib/CIME/get_timing.py
+++ b/scripts/lib/CIME/get_timing.py
@@ -405,7 +405,7 @@ class _TimingParser:
                 comp_label = m.comp + inst_label
             else:
                 comp_label = m.comp
-            self.write("  {} = {:<8s}   {:<6d}      {:<6d}   {:<6d} x {:<6d}  {:<6d} ({:<6d}) \n".format(m.name.lower(), comp_label, (m.ntasks*m.nthrds *smt_factor), m.rootpe, m.ntasks, m.nthrds, m.ninst, m.pstrid))
+            self.write("  {} = {:<8s}   {:<6d}      {:<6d}   {:<6d} x {:<6d}  {:<6d} ({:<6d}) \n".format(m.name.lower(), comp_label, (m.ntasks*m.nthrds), m.rootpe, m.ntasks, m.nthrds, m.ninst, m.pstrid))
             if m.nthrds > maxthrds:
                 maxthrds = m.nthrds
         if self._driver == 'nuopc':
```
